onnxruntime/contrib_ops/cuda/bert/attention_softmax.h

```diff
@@ -179,30 +182,36 @@ __device__ inline void SoftmaxWithRawMaskSmall(const int all_sequence_length,
 
   float thread_data = -CUDART_INF_F;
   if (threadIdx.x < all_sequence_length) {
-    const int batch_index = blockIdx.y;
+    thread_data = float(input[index]) * rsqrt_head_size;
+
     const int sequence_index = blockIdx.x % sequence_length;
+    if (is_unidirectional) {
+      int from_index = all_sequence_length - sequence_length + sequence_index;  // offset of from token in all sequence length.
+      if (threadIdx.x > from_index) {
+        thread_data = -10000.0f;
+      }
+    }
+
     int mask_offset = 0;
+    const int batch_index = blockIdx.y;
     if (mask_dimension == 2) {
       mask_offset = batch_index * all_sequence_length + threadIdx.x;
     } else if (mask_dimension == 3) {
       mask_offset = (batch_index * sequence_length + sequence_index) * all_sequence_length + threadIdx.x;
-    } else if (mask_dimension == 4){
-      // Megatron code:
-      // ltor_mask = ltor_mask[..., (attention_scores.size(3)-hidden_states.size(1)):attention_scores.size(3), :attention_scores.size(3)]
+    } else if (mask_dimension == 4) {
       mask_offset = (batch_index * max_sequence_length + all_sequence_length - sequence_length + sequence_index) * max_sequence_length + threadIdx.x;
```

Comment:
Can we keep the comment here since it's a bit difficult to infer the original logic when looking at the cuda code for mask_dim==4

Reply:
@wangyems, I did not find the code in https://github.com/NVIDIA/Megatron-LM/tree/main/megatron.